text/3691-trusted-publishing-cratesio.md

```diff
@@ -0,0 +1,371 @@
+
+> [!NOTE] 
+> Other security challenges that would persist by not utilizing Trusted Publishing are detailed in the [Motivation](#motivation) section of this document.
+
+## Prior art
+[prior-art]: #prior-art
+
+Other package repositories have implemented similar support:
+
+- [PyPI's Trusted Publishing](https://docs.pypi.org/trusted-publishers/)
+- [RubyGem's Trusted Publishing](https://guides.rubygems.org/trusted-publishing/)
+- [Dart's Automated Publishing](https://dart.dev/tools/pub/automated-publishing)
+- [OSSF's Trusted Publishers For All Package Repositories Blog Post](https://repos.openssf.org/trusted-publishers-for-all-package-repositories)
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Should crate owners be able to configure the allowed token scopes for a Trusted Publisher configuration?
+    - We could default to `publish-new` and `publish-update`, but maybe it's best to allow this to be configurable?
+- How long should an access token derived from the ID token exchange be valid for?
+
+### Out of Scope
+[out-of-scope]: #out-of-scope
+
+- Provenance verification of published crates (e.g. sigstore or other signing mechanisms).
+- Discussion/debate of CI/CD providers to support next and their expected timelines.
+- Support in the GitHub Action used to perform the OIDC authentication with other Rust package registries.
+
+### Items to Resolve Before General Availability
+[resolve-before-general-availability]: #resolve-before-general-availability
+
+- Before crates.io officially supports the OIDC authentication flow from GitHub Actions, we will have an official GitHub Action to facilitate the OIDC token exchange.
+- Crates.io lacks accessible usage documentation. We will need to document the setup of a Trusted Publisher with screenshots and provide example GitHub Action workflows.
+    - Maybe a good spot for this would be in [The Cargo Book](https://doc.rust-lang.org/cargo/index.html)?
+
+## Future possibilities
+[future-possibilities]: #future-possibilities
+
+- Support creating _Trusted Publisher Configurations_ in a `PENDING` state, prior to the initial publishing of a crate.
+- Support for a `DEACTIVATED` state on the Trusted Publisher Configuration to allow toggling on/off the ability to publish crates from the corresponding trusted publisher.
+- Support providing a list of crate names in the crates.io UI to simplify configuration for monorepos that publish multiple crates.
+- Support for additional trusted publishers (i.e. GitLab, CircleCI).
+- Support for custom assertions on OIDC ID token claims.
+    - [Additional claims supported by GitHub Actions OIDC ID tokens](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
```

Comment:
Can we add some discussion of machine-accessible API to provision the configuration here? In particular, for rust-lang/* repositories, I'd love to control this access via rust-lang/team, rather than the current manual configuration process. For some repositories manual configuration (even with this feature) would still mean that we're regularly asking humans to access highly privileged accounts (rust-lang-owner) to update settings, and in a way that likely lacks good review + audit trail (e.g., no 2 person review on the actual policy set).

Reply:
there will definitely be an API for it since that is currently required by the crates.io user interface anyway. I would expect us to only enable it for cookie-based auth at first though until we're happy with the API and consider it stable.

Reply:
I'll add this to the Future Possibilities section. The security around this would be a bit trickier. 🙂